Lib/_pyrepl/fancycompleter.py

```diff
@@ -0,0 +1,175 @@
+                try:
+                    values.append(eval(name, self.namespace))
+                except Exception as exc:
+                    values.append(None)
+        if self.use_colors and names:
+            return self.colorize_matches(names, values)
+        return names
+
+    def attr_matches(self, text):
+        expr, attr = text.rsplit('.', 1)
+        if '(' in expr or ')' in expr:  # don't call functions
+            return []
+        try:
+            thisobject = eval(expr, self.namespace)
+        except Exception:
+            return []
+
+        # get the content of the object, except __builtins__
+        words = set(dir(thisobject)) - {'__builtins__'}
+
+        if hasattr(thisobject, '__class__'):
+            words.add('__class__')
+            words.update(rlcompleter.get_class_members(thisobject.__class__))
+        names = []
+        values = []
+        n = len(attr)
+        if attr == '':
+            noprefix = '_'
+        elif attr == '_':
+            noprefix = '__'
+        else:
+            noprefix = None
+        words = sorted(words)
+        while True:
+            for word in words:
+                if (
+                    word[:n] == attr
+                    and not (noprefix and word[:n+1] == noprefix)
+                ):
+                    try:
+                        val = getattr(thisobject, word)
+                    except Exception:
+                        val = None  # Include even if attribute not set
+
+                    names.append(word)
+                    values.append(val)
+            if names or not noprefix:
+                break
+            if noprefix == '_':
+                noprefix = '__'
+            else:
+                noprefix = None
+
+        if not names:
+            return []
+
+        if len(names) == 1:
+            return [f'{expr}.{names[0]}']  # only option, no coloring.
+
+        prefix = commonprefix(names)
+        if prefix and prefix != attr:
+            return [f'{expr}.{prefix}']  # autocomplete prefix
+
+        if self.use_colors:
+            return self.colorize_matches(names, values)
+
+        if prefix:
+            names += [' ']
+        return names
+
+    def colorize_matches(self, names, values):
+        matches = [self.color_for_obj(i, name, obj)
+                   for i, (name, obj)
+                   in enumerate(zip(names, values))]
+        # We add a space at the end to prevent the automatic completion of the
+        # common prefix, which is the ANSI escape sequence.
+        matches.append(' ')
+        return matches
+
+    def color_for_obj(self, i, name, value):
+        t = type(value)
+        color = self.color_by_type(t)
+        # hack: prepend an (increasing) fake escape sequence,
+        # so that readline can sort the matches correctly.
+        N = f"\x1b[{i:03d};00m"
+        return f"{N}{color}{name}{ANSIColors.RESET}"
+
+    def color_by_type(self, t):
+        theme = get_theme()
```

Comment:
Getting the theme is quite expensive. Creating the theme in `colorize_matches` and passing it to `color_for_obj` and `color_by_type` greatly improves performance (for a large number of matches).

Reply:
fixed in 743e661dc29